src/DIRAC/WorkloadManagementSystem/DB/JobDB.py

```diff
@@ -645,22 +647,27 @@ def setJobAttributes(self, jobID, attrNames, attrValues, update=False, myDate=No
     if not isinstance(jobID, (list, tuple)):
       jobIDList = [jobID]
 
-    jIDList = []
-    for jID in jobIDList:
-      ret = self._escapeString(jID)
-      if not ret['OK']:
-        return ret
-      jIDList.append(ret['Value'])
-
     if len(attrNames) != len(attrValues):
       return S_ERROR('JobDB.setAttributes: incompatible Argument length')
 
     for attrName in attrNames:
       if attrName not in self.jobAttributeNames:
-        return S_ERROR(EWMSSUBM, 'Request to set non-existing job attribute')
+        return S_ERROR(EWMSJERR, 'Request to set non-existing job attribute')
 
     attr = []
+    jobIDListToUpdate = []
+
     for name, value in zip(attrNames, attrValues):
+
+      if name in ('Status', 'MinorStatus', 'ApplicationStatus'):
+        # Need to check what's the current status, and if "final" we don't update
+        for jID in jobIDList:
+          res = self.getJobAttribute(jID, 'Status')
+          if not res['OK']:
+            return S_ERROR(EWMSJERR, "Could not find Status Job Attribute")
+          if res['Value'] not in JobStatus.JOB_FINAL_STATES:
```

Comment:
There is logic in the DataRecoveryAgent that sets Failed jobs to Done or vice versa. This wouldn't work any more, would it?
While it is not strictly necessary to do this, it is a nice cross-check that there are N done jobs and N output files, and sometimes one has to change the Job Status for this purpose.
https://github.com/DIRACGrid/DIRAC/blob/e3aa6a63cafca7a621f45b36e184a63303e6b36c/src/DIRAC/TransformationSystem/Utilities/TransformationInfo.py#L102

Reply:
Actually this would also not work for completed jobs being marked Done once their request is finished?
https://github.com/DIRACGrid/DIRAC/blob/d0dd7dd93bdbc8d72fd84f5cb7dcee90d95fa688/src/DIRAC/WorkloadManagementSystem/Client/JobStatus.py#L61-L63

Reply:
Good review, I was expecting something on these lines. 
For the "Completed -> Done" you are right, I will fix. 

But, for the DataRecoveryAgent logic, I have some doubts. The ultimate role of the DataRecoveryAgent is helping data productions to complete successfully. I think that the scope of action for the DataRecoveryAgent should be "TransformationTasks" and not "Jobs". What we call "ExternalStatus" in the transformation system is, for jobs, equivalent to the DIRAC WMS Job status. I think we should break such equivalence: a job can be "Failed" in WMS while being "Done" in the TS.

Reply:
I CC @phicharp here who might have some good insights.